app/models/csv_file.rb

```diff
@@ -7,4 +11,43 @@ class CSVFile < ApplicationRecord
   as_enum :status, [:created, :processed, :invalid], map: :string,
                                                      source: :status,
                                                      accessor: :whiny
+  delegate :title, to: :dataset
+
+  # mapping path attribute to csv current_path
+  module PathWithCSV
+    def path
+      super || csv.current_path
+    end
+  end
+
+  prepend PathWithCSV
+
+  def convert_to_excel_file
+    excel = Spreadsheet::Workbook.new
+    sheet = excel.create_worksheet name: title
+
+    header_format = Spreadsheet::Format.new(weight: :bold, horizontal_align: :center, locked: true)
+    sheet.row(0).default_format = header_format
+    encoding = CSVParser.new(self).encoding
+
+    CSV.foreach(path, encoding: encoding).with_index do |row, i|
+      sheet.row(i).replace(row)
+    end
+
+    generate_excel_file excel
+  end
+
+  private
+
+  def generate_excel_file(excel)
+    file = Tempfile.new(title)
+
+    begin
+      excel.write file.path
+    ensure
+      file.close
+    end
+
+    file
+  end
```

Comment:
OPTIONAL: redirect user request to a CDN-URI if transfer raw file to user
OPTIONAL: prepare an excel file after csv upload & deliver it to user in the CDN way